File: python/cwe312_storage.py
import logging

logger = logging.getLogger(__name__)


# CWE-312 — Cleartext Storage of Sensitive Information.
def save_credentials(username: str, password: str, api_key: str) -> None:
    # SINK 1 (FILE_WRITE): plaintext password persisted to disk.
    with open("/var/data/creds.txt", "w") as fh:
        fh.write(f"{username}:{password}")

    # SINK 2 (LOG_WRITE): secret written to the log in cleartext.
    logger.info("provisioned user %s with api_key=%s", username, api_key)

File: js/cwe312Storage.js
const fs = require('fs');

// CWE-312 — Cleartext Storage of Sensitive Information.
function saveCredentials(username, password, apiKey) {
  // SINK 1 (FILE_WRITE): plaintext password persisted to a file.
  fs.writeFileSync('/var/data/creds.json', JSON.stringify({ username, password }));

  // SINK 2 (LOG_WRITE): secret written to the console/log in cleartext.
  console.log(`provisioned ${username} with apiKey=${apiKey}`);
}

module.exports = { saveCredentials };
